### ISSUE TITLE
Show a concise summary of keyboard shortcuts when a user hits the "h" key (parity with classic)

### ISSUE THREAD

**Issue by telamonian** [jupyterlab/jupyterlab#10284]
### Feature request

recent ref: #10270

I really miss the simple command cheat sheet from "classic":

![image](https://user-images.githubusercontent.com/2263641/119251581-39c96600-bb75-11eb-8d38-f2674add3ebc.png)

The simple clarity of the "Keyboard shortcuts" modal, plus the ease of bringing it up from within the app with a single <kbd>h</kbd> keypress, is a large part of what got me into IPython Notebook in the first place.

We should bring the "Keyboard shortcuts" modal back in lab. It should be more compact and summarized than eg https://github.com/jupyterlab/jupyterlab-shortcutui, something closer in spirit and design to the original modal from classic.

**Comment by telamonian** [comment]
I do think we should keep the "open help modal on <kbd>h</kbd>" behavior from classic, and I also kind of think that maybe <kbd>h</kbd> should not be remappable (at least in notebook command mode). In general I think it's a good idea to ensure that the help screen is always accessible in the expect way.

@afshin Do we have a precedent for a non-remappable shortcut? Or have we avoided that pattern so far?

**Comment by goanpeca** [comment]
This would indeed help @telamonian. I have a question. Would the idea be to have a modal, per widget (if provided)? I can see this modal being useful if one is in the Notebook, but if I was in the file explorer/terminal, I would expect different shortcut helps to appear.

**Comment by telamonian** [comment]
> I would expect different shortcut helps to appear

That is likely the sticking point that's kept this issue in stasis for the last ~4 years. There is, of course, the currently unused "contextual help". So we could just populate that with context-dependent man-ish pages. Currently, though, "contextual help" opens as a main area widget. This is less than ideal since it will switch away from whatever tab you were actually trying to get help with, as well as completely obscuring it.

So yeah, something like:

"contextual help" but as a sidebar or modal + populating said help screen with a concise summary of the notebook keyboard shortcuts when a notebook is focused.

If that's reasonably successful, we can start thinking about how to populate the help screen in other non-notebook contexts.

**Comment by krassowski** [comment]
+1 for having it as a modal, just like the command palette in 3.0 + having an option to place it in the sidebar instead (just like the command palette). The widget could use a flexbox to show one row in sidebar or two rows when in modal.

**Comment by telamonian** [comment]
> one row

You mean column, right? But yeah, that sounds good

**Comment by fperez** [comment]
I came to comment on this one - the current discoverabilty of the keyboard shortcuts in Lab is really tough. I couldn't remember the one for converting a cell to raw (it's trivially enough `R`;), and I simply wasn't able to find it in Lab i. The docs don't list them, they only direct to the advanced settings editor. But that editor is huge, and finding a single item in it is extremely difficult as its search capabilities are very limited (no jumping to matches, so the search highlights matches but you have to find them by manually scrolling and looking for the highlights).

In the end, I just looked up the classic cheat sheet and trusted it was still the same.  But usability-wise this is pretty awful.  In retrolab it's a scrollable window that I think is also a big UX regression over Classic:

![image](https://user-images.githubusercontent.com/57394/151514526-4e4d5912-5a5b-4799-a82e-2248bb284262.png)

since the classic one is easy to scan at a single glance instead of the "reading a book through a straw" effect of scrolling that much. But at least Retro _has_ the information in an easy-to-find place, whereas for Lab the end-user experience in this regard is pretty rough.

I realize with the more flexible and rich architecture of Lab, a single static sheet model isn't as viable, but I do think this is an important usability regression that I hope we can find a path forward on.

**Comment by jasongrout** [comment]
+1 to getting something up about this. I ran across this frustration the other day, and started thinking: perhaps we could dynamically generate a list of keyboard shortcuts that apply wherever you are currently focused? That would not be quite as good as a curated list and would constantly be changing as you clicked around, or even went in and out of edit/command mode, but it would be something that was kept synced with the current settings.

**Comment by echarles** [comment]
+1 also on this which has been one of my top frustration so far.

>  perhaps we could dynamically generate a list of keyboard shortcuts that apply wherever you are currently focused?

That would be a good feature on top of a simple static immutable base cheatsheet is something very easy to add and what will help the most the user.

**Comment by jasongrout** [comment]
I put up some code at https://github.com/jupyterlab/jupyterlab/pull/11960 that displays relevant keyboard shortcuts in the console for whatever the active element is. If anyone wants to take over that PR and push it forward to display the shortcuts in a dialog or easy popup for the user, go for it.

**Comment by HaudinFlorence** [comment]
> I put up some code at #11960 that displays relevant keyboard shortcuts in the console for whatever the active element is. If anyone wants to take over that PR and push it forward to display the shortcuts in a dialog or easy popup for the user, go for it.

I having a look at it

**Comment by GabrielaVives** [comment]
I have been working on the UX design of this issue, and after a few iterations it seems that these 2 designs stand out : 

<img width="1398" alt="image" src="https://user-images.githubusercontent.com/105775162/221612498-272e7460-e439-4799-af5f-5e2842fa4be7.png">

The UX improvements are (that these 2 designs have in common):
- Removing the titles of the columns (no need to have 'label' and 'tooltips', it is already explicit)
- Removing the "dismiss" button and replacing it by a "close" icon at the top right, this liberates the blank space at the bottom
- Putting the labels to the left of the shortcuts, as in a 'reading from left to right' logic
- having a rectangle around each key (imitating the how a keyboard key physically looks like) instead of it being around the whole combination
- Adding padding of at least 14 px between the columns and at least 8px between each line

Some questions still remain, and the above A and B designs illustrate to solutions to the following problem: shortcuts may be made of several combinations, for example (CTRL+T followed by CTRL+P) = 1 shortcut

_Solution A_
-  '+' in between each key
- a blank space in between 2 combinations

_Solution B_
- a blank space in between each key
- '+' between 2 combinations

I think that these 2 solutions work great from a usability point of view, I would love to hear you opinion on them.

**Comment by krassowski** [comment]
> Some questions still remain, and the above A and B designs illustrate to solutions to the following problem: shortcuts may be made of several combinations, for example (CTRL+T followed by CTRL+P) = 1 shortcut

Elsewhere in the UI and in the old Notebook comma (`,`) signifies sequence of keys, e.g. `D, D` in:

![Screenshot from 2023-02-27 16-32-58](https://user-images.githubusercontent.com/5832902/221623012-1fe01ca0-457d-4638-85ed-01a6e883aea5.png)

Also, for inspiration the GNOME way (grouping shortcuts into categories):

![Screenshot from 2023-02-27 16-30-08](https://user-images.githubusercontent.com/5832902/221622254-1612d3a7-6d5a-4a1b-9763-783b2f957289.png)

> I think that these 2 solutions work great from a usability point of view, I would love to hear you opinion on them.

Old notebook has a very prominent command/edit mode exposure. Can we replicate this somehow? It is a key lesson on using notebooks efficiently.

**Comment by HaudinFlorence** [comment]
@krassowski Thanks for your input. I am not sure to fully understand what your last comment may induce on the design.

**Comment by GabrielaVives** [comment]
> Elsewhere in the UI and in the old Notebook comma (`,`) signifies sequence of keys, e.g. `D, D` in:

I thought about that, but finally decided there was no need, as when you duplicate the same key, it is obvious that you need to press it twice.

 For example here: 
 
<img width="284" alt="image" src="https://user-images.githubusercontent.com/105775162/221626805-0de17e98-0e3d-411a-8749-77cf398356f5.png">

Also, for inspiration the GNOME way (grouping shortcuts into categories):

If you have some categories in mind, that would be great! I had done this design for categories, but I do not know what the best organisation of the categories would be

<img width="677" alt="image" src="https://user-images.githubusercontent.com/105775162/221627337-8eee1bfb-de79-4be4-b235-249e2ccc852a.png">

**Comment by HaudinFlorence** [comment]
> I have been working on the UX design of this issue, and after a few iterations it seems that these 2 designs stand out :
> 
> <img alt="image" width="1398" src="https://user-images.githubusercontent.com/105775162/221612498-272e7460-e439-4799-af5f-5e2842fa4be7.png">
> 
> The UX improvements are (that these 2 designs have in common):
> 
> * Removing the titles of the columns (no need to have 'label' and 'tooltips', it is already explicit)
> * Removing the "dismiss" button and replacing it by a "close" icon at the top right, this liberates the blank space at the bottom
> * Putting the labels to the left of the shortcuts, as in a 'reading from left to right' logic
> * having a rectangle around each key (imitating the how a keyboard key physically looks like) instead of it being around the whole combination
> * Adding padding of at least 14 px between the columns and at least 8px between each line
> 
> Some questions still remain, and the above A and B designs illustrate to solutions to the following problem: shortcuts may be made of several combinations, for example (CTRL+T followed by CTRL+P) = 1 shortcut
> 
> _Solution A_
> 
> * '+' in between each key
> * a blank space in between 2 combinations
> 
> _Solution B_
> 
> * a blank space in between each key
> * '+' between 2 combinations
> 
> I think that these 2 solutions work great from a usability point of view, I would love to hear you opinion on them.

@GabrielaVives I am sharing a screen capture of the current state of the shortcut display close to your A prototype
![Screenshot from 2023-03-07 15-27-29](https://user-images.githubusercontent.com/99649086/223452102-9c90f26d-93ec-4fa7-8e07-cd887a47c9fc.png)

The point "Removing the "dismiss" button and replacing it by a "close" icon at the top right" to leberate the blank space at the bottom, is not so easy to change since the shortcuts displays are shown in a dialog display having its own style that can be changed independently.

**Comment by andrii-i** [comment]
I agree with @fperez and @krassowski that the best solution would be to show table in a way that all shortcuts would be visible at the same time and there would be no scrolling. Next best solution would be to have some scrolling but still have multiple rows to use space more efficiently. Finally having any shortcut table beats not having shortcuts table at all.

**Comment by GabrielaVives** [comment]
@HaudinFlorence it seems that the rows are slightly misaligned, making it difficult to concentrate and find the correct shortcut. 
I wonder if it is because the text on the left column is aligned to the top? Whereas aligning it in the middle would make it easier to follow the line. Also, maybe increasing the spacing between the lines would help.

Here are my recommendations:

- Align the text in the middle of the row
- Padding around the “+” equivalent to 1 space
- Increased spacing between the lines (between 3 to 6 px, needs to be tried out)
- Maybe align the text to the right? here is an illustration of the 2 solutions:

<img width="802" alt="image" src="https://user-images.githubusercontent.com/105775162/223658561-0776739f-a33b-4037-aff1-e4ad61554c7c.png">


Concerning the size of the modal, is it possible to make it longer (to see more information with less scrolling), or is it tied to the modal's style?

No problem about the 'Close' button, I think it is also a good solution.

@andrii-i @fperez @krassowski would you be interested (and any other users who may be interested) in having a session to find the best groups for these shortcuts? It would be in a workshop format of less than an hour, as an outcome we would have a first version of the information architecture of this modal.

**Comment by HaudinFlorence** [comment]
@andrii-i @GabrielaVives Thanks for your comments and sorry for the vertical misalignment issue. Here a better styling (hopefully). I would be happy to discuss the shortcut display panel during a session too.

![Screenshot from 2023-03-08 14-22-22](https://user-images.githubusercontent.com/99649086/223724221-86947afa-3f1c-41a8-bb5b-0349b68f5e96.png)

**Comment by HaudinFlorence** [comment]
> I agree with @fperez that the best solution would be to show table in a way that all shortcuts would be visible at the same time and there would be no scrolling. Next best solution would be to have some scrolling but still have multiple rows to use space more efficiently (something like GNOME screenshot shared by @krassowski but without named sections). Finally having any shortcut table beats not having shortcuts table at all.

@andrii-i I can try to do something about having several columns instead of having to scroll.

**Comment by GabrielaVives** [comment]
@HaudinFlorence yes it's clearer now. I think it's looking good ! Yes, let's organise that :) Maybe the version with the columns and no scrolling could be a V2.

**Comment by jasongrout** [comment]
I commented more extensively [here](https://github.com/jupyterlab/jupyterlab/pull/14053#issuecomment-1460406214), but I'll also note on this issue since it seems that design work is being done here: we tried hard to follow platform conventions when we display keyboard shortcuts elsewhere in the UI. For example, on mac keys are separated by spaces (not `+`), and symbols represent the various modifiers. We encapsulate this platform-specific logic in the CommandRegistry [formatKeystroke function](https://github.com/jupyterlab/lumino/blob/06a87ead9b6b87842e8592fda8a15f27802c0646/packages/commands/src/index.ts#L1206) in Lumino, which I wrapped in the original PR [here](https://github.com/jupyterlab/jupyterlab/pull/11960/files#diff-bff7008237b575f8298c99c905c175a510873cb9a948ebc9c6226a7cfb252bd1R48-R50). I'd strongly encourage us to (a) be consistent with other places in JLab that we display shortcuts (like menus, etc.), and (b) be consistent with platform conventions (e.g., `+` separates keys on non-mac, space separates keys on mac).

**Comment by fperez** [comment]
> @andrii-i @fperez @krassowski would you be interested (and any other users who may be interested) in having a session to find the best groups for these shortcuts? It would be in a workshop format of less than an hour, as an outcome we would have a first version of the information architecture of this modal.

Thanks for the great work @GabrielaVives! I'd love to be able to but unfortunately right now I'm swamped and can't, sorry! But I am happy to see this progress, it will be a big usability win for many.

**Comment by andrii-i** [comment]
@GabrielaVives @HaudinFlorence  Please also see my more detailed [comment](https://github.com/jupyterlab/jupyterlab/pull/14053#issuecomment-1459060634) on the subject in the PR #14053 itself (lots of screenshots), I would also recommend to have all the discussion on the topic in single thread

**Comment by jasongrout** [comment]
FYI, the original PR grouped the shortcuts by specificity. In other words, shortcuts that applied most specifically to your context are listed first, application shortcuts are listed last. Here's a screenshot showing the space separating two groups:

<img width="1025" alt="Screen Shot 2023-03-08 at 10 47 35" src="https://user-images.githubusercontent.com/192614/223791584-c702c2b0-2198-4b88-a2ec-c95743e3e869.png">

IIRC, this was implemented technically by walking up the DOM elements to the root, and at each element asking for shortcuts applied to that element. Then the shortcuts are grouped by which ancestor they applied to, as a proxy for relevance to your context.

There aren't labels for these divisions, and I'm not sure how we could easily get labels.

Of course, another way to group shortcuts is by the command category, which would be similar to how commands were grouped in the original command palette. This would be easy since we have a names for these divisions that we used for the command palette at one point.

**Comment by andrii-i** [comment]
> @andrii-i @GabrielaVivnks for your comments and sorry for the vertical misalignment issue. Here a better styling (hopefully). I would be happy to discuss the shortcut display panel during a session too.

@HaudinFlorence This option looks much more readable to me. I think making text in Labels column right-aligned as in A.2 preview by @GabrielaVives above would increase readability even more

**Comment by GabrielaVives** [comment]
**Workshop Update**

As discussed during the JupyterLab meeting yesterday, a workshop to talk about the V2 of this feature will be held on March 16th, at 5:30pm (CET). There will be a technical discussion as well as a UX workshop. 

I updated the [meeting notes](https://hackmd.io/@ajbozarth/H13kVN5RI), where you can find the link to the workshop, as well as details in the description. This workshop is open to anyone interested in participating or just listening in.

**Comment by jasongrout** [comment]
Thanks for scheduling this! How long is the scheduled meeting? 1 hour?

For convenience to others, here is your local time:
https://arewemeetingyet.com/Paris/2023-03-16/17:30

Thanks,

Jason

**Comment by GabrielaVives** [comment]
@jasongrout it is scheduled for 1h30. Sorry I only put the description on the calendar notes but I'll post them here as well.

@andrii-i Yes for sure we should put it on to the Jupyter Calendar, I'll update my previous comment with the meeting description.

**Comment by andrii-i** [comment]
@GabrielaVives sounds good, please let us know when updating meeting description update is done, personally I don't have calendar edit access but I can ask my team to create it

**Comment by jasongrout** [comment]
I added an event to the Jupyter calendar (please check to make sure it is correct). Do you have a videoconference link we can add as well? We can't use the normal Jupyter zoom channel since another Jupyter meeting will be using it.

**Comment by JasonWeill** [comment]
We can meet using the Jovyan Zoom room: https://zoom.us/my/jovyan?pwd=c0JZTHlNdS9Sek9vdzR3aTJ4SzFTQT09

Triage normally takes place 10:00–11:00 PDT, but we can skip that next week in favor of this discussion. In addition, we can shorten the Jupyter Server contributing hour to 30 minutes next week.

**Comment by GabrielaVives** [comment]
@JasonWeill , ok let's use this room. I'll update the link in the meeting description as well.

**Comment by andrii-i** [comment]
@GabrielaVives @HaudinFlorence congrats on closing this, good job! Wanted to ask if you think #14240 is enough for tracking Keyboard Shortcuts V2 / improvements to V1 or if separate issue should be created for that?

**Comment by GabrielaVives** [comment]
Maybe we should create separate issues for each item as we tackle them?
